### PR TITLE
SCIM: Keep user emails in sync with SCIM patch and replace requests

### PR DESCRIPTION
**Pull request by chwarwick**
This adds logic to ensure the user_emails table in updated in response to the PATCH and PUT requests that modify emails.

closes https://github.com/sourcegraph/sourcegraph/issues/48720

Notes: 

- Currently the create endpoint does not enforce that all emails are successfully created, it only enforces that "primary" emails are created.  We may want to reconsider this decision as it can make the SCIM data and the emails table out of sync.
- The PATCH/PUT endpoints **do** enforce that all requested complete successfully however they do also account to the possibility that emails may exist that SCIM doesn't know about and will reuse these. 
- The PATCH/PUT endpoints **will not** remove any additional emails that were not added via SCIM.  These remain in the table and can be successfully added via future SCIM requests. Once SCIM is aware of the emails they can be removed in future requests.
- SCIM allows multiple primary emails, however Sourcegraph does not in the case of multiple primaries we take the first one.  

## Test plan
Unit tests add 

Test the following operations manually all successful except where noted
- PATCH - Add a new email and make it primary
- PATCH - Add an email that already exists for another user but is currently not verified
- PATCH - Make email from the previous step primary
- PATCH - Add a non-primary email that already exists for another user that was already verified
  -  This request fails correctly because once an email is verified it must be unique
- PUT - Replace email with a new one
- PUT - Replace email with an email that already existed on the user but not from SCIM
- PUT - Replace and don't provide any email addresses
  - This request fails because it is not valid to remove the primary email

### CODE REVIEW

File: enterprise/internal/scim/user_patch.go [hunk 1]
```diff
@@ -45,6 +47,7 @@ func (h *UserResourceHandler) Patch(r *http.Request, id string, operations []sci
 				for rawPath, value := range op.Value.(map[string]interface{}) {
 					newlyChanged := applyChangeToAttributes(userRes.Attributes, rawPath, value)
 					changed = changed || newlyChanged
+					emailsModified = true
```
Comment (chwarwick) [inline enterprise/internal/scim/user_patch.go:50]: This case accounts for changes to the empty path, so just assume the emails are changed.  A false positive here is not problematic.

File: enterprise/internal/scim/user.go [hunk 2]
```diff
@@ -116,9 +120,119 @@ func updateUser(ctx context.Context, tx database.DB, oldUser *types.UserForSCIM,
...
 }
 
+type emailUpdates struct {
+	toRemove       []string
+	toAdd          []string
+	toVerify       []string
+	resetPrimaryTo *string
+}
+
+func generateEmailUpdates(startingUserData, endingUserData scim.ResourceAttributes, currentEmails []*database.UserEmail) emailUpdates {
```
Comment (vdavid) [inline enterprise/internal/scim/user.go:178]: The names `startingUserData` and `endingUserData` are not intuitive to me, as in, I didn't understand from the code what they meant.
How about either
a.) renaming the function to `diffEmails`, and the arguments to `a` and `b`, and also maybe `currentEmails` to `emailsInDB`, or
b.) adding some docs over the function?

File: enterprise/internal/scim/user.go [hunk 3]
```diff
@@ -116,9 +120,119 @@ func updateUser(ctx context.Context, tx database.DB, oldUser *types.UserForSCIM,
...
+	startingPrimary, startingOther := extractPrimaryEmail(startingUserData)
+	endingPrimary, endingOther := extractPrimaryEmail(endingUserData)
```
Comment (vdavid) [inline enterprise/internal/scim/user.go:180]: Another naming tip: for `*Other`, I'd use `*Others` to document that they are arrays.

File: enterprise/internal/scim/user.go [hunk 4]
```diff
@@ -116,9 +120,119 @@ func updateUser(ctx context.Context, tx database.DB, oldUser *types.UserForSCIM,
...
+				return err
+			}
+			err = tx.UserEmails().SetVerified(ctx, oldUser.ID, newEmail, true)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Now verify any addresses that already existed and weren't verified
+		for _, verifyEmail := range updates.toVerify {
```
Comment (vdavid) [inline enterprise/internal/scim/user.go:142]: Nit: I'd call the loop variable just `email`, I think it's more descriptive of what it is.
(I'd probably do the same for the loop variables in all three loops.)

File: enterprise/internal/scim/user.go [hunk 5]
```diff
@@ -116,9 +120,119 @@ func updateUser(ctx context.Context, tx database.DB, oldUser *types.UserForSCIM,
...
+		// or we required the add to succeed in the prior steps
+		if updates.resetPrimaryTo != nil {
+			err = tx.UserEmails().SetPrimaryEmail(ctx, oldUser.ID, *updates.resetPrimaryTo)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Finally remove any email addresses
+		for _, newEmail := range updates.toRemove {
```
Comment (vdavid) [inline enterprise/internal/scim/user.go:160]: Probably copy-paste mistake here to call this "new"

File: enterprise/internal/scim/user.go [hunk 6]
```diff
@@ -116,9 +120,119 @@ func updateUser(ctx context.Context, tx database.DB, oldUser *types.UserForSCIM,
 		return scimerrors.ScimError{Status: http.StatusInternalServerError, Detail: errors.Wrap(err, "could not update").Error()}
 	}
 
+	if emailsModified {
+		currentEmails, err := tx.UserEmails().ListByUser(ctx, database.UserEmailsListOptions{UserID: oldUser.ID, OnlyVerified: false})
+		if err != nil {
+			return err
+		}
+		updates := generateEmailUpdates(startingUserSCIMResource.Attributes, attributes, currentEmails)
```
Comment (vdavid) [inline enterprise/internal/scim/user.go:128]: I like the approach of first diffing the attributes and then applying the diff to the DB! I think it's very easy to grasp, and breaks down the complexity really well.
Just a thought to further improve things: I'd try to capture the "diff → apply diff" terminology more, perhaps in the naming of the functions, variable, and struct, or in comments.

**Review by vdavid — APPROVED** [review]
> Currently the create endpoint does not enforce that all emails are successfully created, it only enforces that "primary" emails are created. We may want to reconsider this decision as it can make the SCIM data and the emails table out of sync.

I think it's good enough for v1. I think of the `account_data` field as the source of truth for SCIM-controlled users (as it's a verbatim copy of the data that arrived from the IdP which is the actual source of truth)—maintaining our own fields is like a side-effect of the attributes being set, and I don't mind if that representation is not perfect in some edge cases.

> The PATCH/PUT endpoints do enforce that all requested complete successfully however they do also account to the possibility that emails may exist that SCIM doesn't know about and will reuse these.
> The PATCH/PUT endpoints will not remove any additional emails that were not added via SCIM. These remain in the table and can be successfully added via future SCIM requests. Once SCIM is aware of the emails they can be removed in future requests.

I could argue that PUT should remove all emails not listed in the request, but it's low stakes, I don't mind it now. This implementation seems good enough!

> SCIM allows multiple primary emails, however Sourcegraph does not in the case of multiple primaries we take the first one.
👍

**Comment by chwarwick** [comment]
> I could argue that PUT should remove all emails not listed in the request, but it's low stakes, I don't mind it now. This implementation seems good enough!
> 
> >

I debated that one too and it's an easy change to make.  What I wasn't sure of is if there were other complications that it could cause if there were additional external accounts that had emails associated with them for permissioning purposes to code hosts.  That is why I left them in for now.

**Comment by chwarwick** [comment]
> I think of the `account_data` field as the source of truth for SCIM-controlled users (as it's a verbatim copy of the data that arrived from the IdP which is the actual source of truth)—maintaining our own fields is like a side-effect of the attributes being set, and I don't mind if that representation is not perfect in some edge cases.

This is the part that I find potentially concerning, aside from the scim resource handler to my knowledge no other part of sg knows anything about the account_data fields and only look at the other database values.

**Comment by vdavid** [comment]
> This is the part that I find potentially concerning, aside from the scim resource handler to my knowledge no other part of sg knows anything about the account_data fields and only look at the other database values.

We can later:
- Change the source of truth to `account_data`
- Move `account_data` to a different structure entirely, maybe to a `user_scim_entries` or similar table to make the imported data more structured. `user_external_accounts` was originally for a different purpose as you know, and although it was luckily flexible enough to fit this data in there, it's not specific enough to store this data in an easy-to-query way

Till then, the most we can do is to ensure that we consistently mirror the data in other fields.

**Pull request merged**